core/blockchain/impl/block_storage_impl.cpp

```diff
@@ -224,92 +223,101 @@ namespace kagome::blockchain {
+      const primitives::Justification &justification,
+      const primitives::BlockHash &hash) {
+    BOOST_ASSERT(not justification.data.empty());
 
-    OUTCOME_TRY(justification, scale::encode(j));
+    OUTCOME_TRY(encoded_justification, scale::encode(justification));
     OUTCOME_TRY(putToSpace(*storage_,
                            Space::kJustification,
-                           block_number,
                            hash,
-                           Buffer{justification}));
+                           std::move(encoded_justification)));
 
     // the following is still required
     primitives::BlockData block_data{.hash = hash};
-    OUTCOME_TRY(putBlockData(block_number, block_data));
+    OUTCOME_TRY(putBlockData(block_data));
+
     return outcome::success();
   }
 
   outcome::result<void> BlockStorageImpl::removeJustification(
-      const primitives::BlockHash &hash, primitives::BlockNumber number) {
-    auto key = numberAndHashToLookupKey(number, hash);
-    auto space = storage_->getSpace(Space::kJustification);
+      const primitives::BlockHash &block_hash) {
+    const auto &key = block_hash;
 
+    auto space = storage_->getSpace(Space::kJustification);
     OUTCOME_TRY(space->remove(key));
+
     return outcome::success();
   }
 
   outcome::result<void> BlockStorageImpl::removeBlock(
-      const primitives::BlockInfo &block) {
-    auto block_lookup_key = numberAndHashToLookupKey(block.number, block.hash);
+      const primitives::BlockHash &block_hash) {
+    // Check if block still in storage
+    OUTCOME_TRY(header_opt, getBlockHeader(block_hash));
+    if (not header_opt.has_value()) {
+      return outcome::success();
+    }
+    const auto &header = header_opt.value();
 
-    SL_TRACE(logger_, "Removing block {}...", block);
-    auto key_space = storage_->getSpace(Space::kLookupKey);
+    primitives::BlockInfo block_info(header.number, block_hash);
 
-    if (auto res = key_space->remove(block.hash); res.has_error()) {
-      logger_->error("could not remove hash-to-idx from the storage: {}",
-                     res.error());
-      return res;
-    }
+    SL_TRACE(logger_, "Removing block {}...", block_info);
 
-    auto num_to_idx_key = numberToIndexKey(block.number);
-    OUTCOME_TRY(num_to_idx_val_opt, key_space->tryGet(num_to_idx_key.view()));
-    if (num_to_idx_val_opt == block_lookup_key) {
-      if (auto res = key_space->remove(num_to_idx_key); res.has_error()) {
-        SL_ERROR(logger_,
-                 "could not remove num-to-idx from the storage: {}",
-                 block,
-                 res.error());
-        return res;
+    {  // Remove number-to-key assigning
+      auto num_to_hash_key = blockNumberToKey(block_info.number);
+
+      auto key_space = storage_->getSpace(Space::kLookupKey);
+      OUTCOME_TRY(hash_opt, key_space->tryGet(num_to_hash_key.view()));
+      if (hash_opt == block_hash) {
+        if (auto res = key_space->remove(num_to_hash_key); res.has_error()) {
```

Comment:
Do we need `std::mutex db_num_to_hash_mutex` to ensure that value didn't change between `tryGet` and `remove`?

Reply:
It should be covered by future db-tx